types/wlr_input_timestamps_v1.c

```diff
@@ -0,0 +1,383 @@
+	wlr_input_timestamps_manager_v1_destroy(manager);
+}
+
+/**
+ * Implementations
+ */
+
+static const struct zwp_input_timestamps_manager_v1_interface manager_impl = {
+	.destroy = manager_handle_destroy,
+	.get_keyboard_timestamps = manager_handle_get_keyboard_timestamps,
+	.get_pointer_timestamps = manager_handle_get_pointer_timestamps,
+	.get_touch_timestamps = manager_handle_get_touch_timestamps,
+};
+
+static const struct zwp_input_timestamps_v1_interface impl = {
+	.destroy = handle_destroy,
+};
+
+/**
+ * Public functions
+ */
+
+struct wlr_input_timestamps_manager_v1 *wlr_input_timestamps_manager_v1_create(struct wl_display *display) {
+	struct wlr_input_timestamps_manager_v1 *manager =
+		calloc(1, sizeof(struct wlr_input_timestamps_manager_v1));
+
+	if (manager == NULL) {
+		return NULL;
+	}
+
+	wl_list_init(&manager->resources);
+	wl_list_init(&manager->input_timestamps);
+
+	manager->global = wl_global_create(display,
+		&zwp_input_timestamps_manager_v1_interface, INPUT_TIMESTAMPS_MANAGER_VERSION,
+		manager, manager_bind);
+
+	if (manager->global == NULL) {
+		free(manager);
+		return NULL;
+	}
+
+	wl_signal_init(&manager->events.destroy);
+	wl_signal_init(&manager->events.new_keyboard_timestamps);
+	wl_signal_init(&manager->events.new_pointer_timestamps);
+	wl_signal_init(&manager->events.new_touch_timestamps);
+
+	manager->display_destroy_listener.notify = manager_handle_display_destroy;
+	wl_display_add_destroy_listener(display, &manager->display_destroy_listener);
+
+	wlr_log(WLR_DEBUG, "input_timestamps_v1 manager created");
+
+	return manager;
+}
+
+void wlr_input_timestamps_manager_v1_destroy(struct wlr_input_timestamps_manager_v1 *manager) {
+	if (manager == NULL) {
+		return;
+	}
+
+	wlr_signal_emit_safe(&manager->events.destroy, manager);
+	wl_list_remove(&manager->display_destroy_listener.link);
+
+	struct wlr_input_timestamps_v1 *object, *tmp_object;
+	wl_list_for_each_safe(object, tmp_object, &manager->input_timestamps, link) {
+		wl_resource_destroy(object->resource);
+	}
+
+	struct wl_resource *resource, *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &manager->resources) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_global_destroy(manager->global);
+	free(manager);
+}
+
+/**
+ * Sends input event timestamps in nanosecond accuracy depending on
+ * input_timestamps object type.
+ */
+
+static void manager_send_timestamp(struct wlr_input_timestamps_manager_v1 *manager,
+		struct wlr_seat *seat, uint64_t tv_sec, uint32_t tv_nsec,
+		enum wlr_input_timestamps_type input_timestamps_type) {
+	struct wlr_seat_client *focused_client = NULL;
+	switch (input_timestamps_type) {
+	case WLR_INPUT_TIMESTAMPS_KEYBOARD:;
+		focused_client = seat->keyboard_state.focused_client;
+		break;
+	case WLR_INPUT_TIMESTAMPS_POINTER:;
+		focused_client = seat->pointer_state.focused_client;
+		break;
+	case WLR_INPUT_TIMESTAMPS_TOUCH:;
+		// TODO wlr_seat_touch_state has no focused_client member
```

Comment:
Hmm, this is because there can be multiple touches, each on a different surface. Could be left as a follow-up issue.

I'd prefer doing `focused_client = NULL` instead of sending events to an unrelated client though.

Reply:
Setting `focused_client = NULL` makes it ignore this call for touch events because of the following check. It should ideally be send to all clients receiving the corresponding event; let me check how this stuff is handled elsewhere.

Reply:
>Setting focused_client = NULL makes it ignore this call for touch events because of the following check.

Yes, that's intended.

>It should ideally be send to all clients receiving the corresponding event; let me check how this stuff is handled elsewhere.

No, only to the client receiving the touch point. A new `touch_point` argument is likely required.